### DISCUSSION TITLE
How can I share a site with external user with pnpcore

### DISCUSSION THREAD

**Discussion by pmatthews05** [pnp/pnpcore#785] [Q&A]
I'm going through update code from SharePointPNPCoreOnline v3.7.1903 and re-writing everything using PnP.Core.

I'm unable to find anything in the documentation on how to share.

In my previous code I did this two different ways. Using standard Microsoft.SharePoint.Client.Sharing
```c#
var srs = WebSharingManager.UpdateWebSharingInformation(clientContext, clientContext.Web, users, sendEmail, message, true, true);
await clientContext.ExecuteQueryRetryAsync();
```
Also using SharePointPNPCoreOnline
```c#
var visitorsGroup = clientContext.Web.AssociatedVisitorGroup;
clientContext.Load(visitorsGroup, g => g.Id);
await clientContext.ExecuteQueryRetryAsync();
SharingResult sr = await visitorsGroup.InviteExternalUserAsync(externalAccessItem.GuestEmail, sendEmail, message);
```
Could someone point me in the right direction how to do this with the new PnP.Core package please?

**Comment by magnotonus** [comment]
Hi, i'm running the same issue. I Can't found how to share with external users using the new package and had to use old components just for this.
Any idea?

**Reply by pmatthews05** [reply]
I've think I've found a work around (for now), needing to look at this later today to see if it fully works. But using MSGraph I can send an 'Invitation' to the external user and include the siteURL.

As soon at that is sent I can then ensure the user in the site and add them as a user. I've only tested this so far when the user is already an external user in our AD. It might be if the user never been invited before, they are not available to SharePoint until they have accepted the invite.

My code is as follows:
dotnet add package Microsoft.Graph
```csharp
using Microsoft.Graph

 public async Task SendInviteToSite(string externalUserEmail, string siteUrl)
        {
            _log.LogInformation($"{LOG_LOCATION}:SendInviteToSite:Connecting to 'Default' SharePoint...");
            using (var pnpCoreContext = await _pnpContextFactory.CreateAsync("Default"))
            {
                _log.LogInformation($"{LOG_LOCATION}:SendInviteToSite:Connecting to Graph service...");
                var graphServiceClient = new GraphServiceClient(new DelegateAuthenticationProvider((requestMessage) =>
                {
                    return pnpCoreContext.AuthenticationProvider.AuthenticateRequestAsync(new Uri("https://graph.microsoft.com"), requestMessage);
                }));

                var invitation = new Invitation
                {
                    InvitedUserEmailAddress = item.Title,
                    InviteRedirectUrl = $"https://{Environment.GetEnvironmentVariable("Tenant")}.sharepoint.com{siteUrl}",
                    SendInvitationMessage = true,
                    InvitedUserMessageInfo = new InvitedUserMessageInfo
                    {
                        CustomizedMessageBody = Environment.GetEnvironmentVariable("WelcomeMessage")
                    }
                };
                _log.LogInformation($"{LOG_LOCATION}:SendInviteToSite:Sending invitation for {externalUserEmail} for site {siteUrl}...");
                await graphServiceClient.Invitations.Request().AddAsync(invitation);
            }
        }
```
Then i can add the person to the Visitors Group.
```csharp
 private async Task GrantVisitorPermission(string site, string externalUserEmail)
        {
            _log.LogInformation($"{LOG_LOCATION}:GrantVisitorPermission:Connecting to SharePoint...");
            var siteUrl = $"https://{Environment.GetEnvironmentVariable("Tenant")}.sharepoint.com{site}";
            using (var pnpContext = _pnpContextFactory.Create(new Uri(siteUrl)))
            {
                await pnpContext.Web.LoadAsync(p => p.AssociatedVisitorGroup);
                var user = await pnpContext.Web.EnsureUserAsync(externalUserEmail);
                await pnpContext.Web.AssociatedVisitorGroup.AddUserAsync(user.LoginName);
                _log.LogInformation($"{LOG_LOCATION}:GrantVisitorPermission:Added {user.Title} to site {site} as Visitor");
            }
        }
```

**Reply by magnotonus** [reply]
Great! It works.
I confirm that after the first step is executed, an email is sent to the email address with the invitation. After it is accepted we can run the second step.

We will use your solution meanwhile found another more direct way, but it is something at least.

Thanks pmatthews05 !!

**Reply by magnotonus** [reply]
Update:
It is possible to automaticaly accept the invitation with just a few changes:

    public async Task SendInviteToSite(string externalUserEmail, string siteUrl)
        {
            _log.LogInformation($"{LOG_LOCATION}:SendInviteToSite:Connecting to 'Default' SharePoint...");
            using (var pnpCoreContext = await _pnpContextFactory.CreateAsync("Default"))
            {
                _log.LogInformation($"{LOG_LOCATION}:SendInviteToSite:Connecting to Graph service...");
                var graphServiceClient = new GraphServiceClient(new DelegateAuthenticationProvider((requestMessage) =>
                {
                    return pnpCoreContext.AuthenticationProvider.AuthenticateRequestAsync(new Uri("https://graph.microsoft.com"), requestMessage);
                }));

                var invitation = new Invitation
                {
                    InvitedUserEmailAddress = item.Title,
                    InviteRedirectUrl = $"https://{Environment.GetEnvironmentVariable("Tenant")}.sharepoint.com{siteUrl}",
                    SendInvitationMessage = false, // ******** do not send the email invitation ************
                    InvitedUserMessageInfo = new InvitedUserMessageInfo
                    {
                        CustomizedMessageBody = Environment.GetEnvironmentVariable("WelcomeMessage")
                    }
                };
                _log.LogInformation($"{LOG_LOCATION}:SendInviteToSite:Sending invitation for {externalUserEmail} for site {siteUrl}...");
                var _inv = await graphClient.Invitations.Request().AddAsync(invitation); // ******** assign invitation result to an object *********
                _inv.Status = "Completed"; // ********* change invitation status *************
            }
        }


So you can immediately run the second step.
Hope this to be useful.
